documentation/design_docs/vpn/vpn_role_interface_design.md

```diff
@@ -0,0 +1,334 @@
+```yaml
+all:
+  hosts:
+    bastion1.example.com:
+      vpn_hostid: '@bastion1'
+      vpn_shared_key: Vault!abc..  #  vault encrypted value
+      vpn_connections:
+        - name: bastion1-bastion2
+          remote: bastion2.example.com
+    bastion2.example.com:
+      vpn_hostid: '@bastion2'
+      vpn_shared_key: Vault!def..  #  vault encrypted value
+      vpn_connections:
+        - name: bastion2-bastion1
+          remote: bastion1.example.com
+        - name: bastion2-bastion3
+          remote: bastion3.example.com
+    bastion3.example.com:
+      vpn_hostid: '@bastion3'
+      vpn_shared_key: Vault!ghi..  #  vault encrypted value
+      vpn_connections:
+        - name: bastion3-bastion2
+          remote: bastion2.example.com
+  children:
+    vpn:
+      bastion1.example.com:
+      bastion2.example.com:
+      bastion3.example.com:
+```
+And I'll use a playbook like this:
+```yaml
+- hosts: vpn
+  roles:
+    - linux-system-roles.vpn
+```
+We can use `hostvars` to look up the parameters of the other hosts:
+```
+set remotehost = vpn_connections[index].remote  #  e.g. bastion2.example.com
+set remotevars = hostvars[remotehost]
+conn {{ vpn_connections[index].name }}
+    left={{ inventory_hostname }}
+    leftid={{ vpn_hostid }}
+    right={{ remotehost }}
+    rightid={{ remotevars.remoteid }}
+# and the shared key comes from remotevars.shared_key
+```
+We could also define `vpn_connections` at the global level, but we would have to
+add some sort of logic to the template so that it would skip setting up tunnels
+to itself (e.g. when `inventory_hostname == remotehost`)
+
+If we wanted to do a full mesh configuration, we would omit the
+`vpn_connections`, and include the logic to skip setting up tunnels to itself.
+```yaml
+all:
+  hosts:
+    bastion1.example.com:
+      vpn_hostid: '@bastion1'
+      vpn_shared_key: Vault!abc..  #  vault encrypted value
+    bastion2.example.com:
+      vpn_hostid: '@bastion2'
+      vpn_shared_key: Vault!def..  #  vault encrypted value
+    bastion3.example.com:
+      vpn_hostid: '@bastion3'
+      vpn_shared_key: Vault!ghi..  #  vault encrypted value
+  children:
+    vpn:
+      bastion1.example.com:
+      bastion2.example.com:
+      bastion3.example.com:
+```
+And I'll use a playbook like this:
+```yaml
+- hosts: vpn
+  roles:
+    - linux-system-roles.vpn
+```
+The vpn role would be called for each host in the group. The vpn role code would
+loop over all hosts defined in the current play, excluding the current host, and
+set up vpn tunnels to all of them:
+```
+for host in all_hosts | reject('==', inventory_hostname)
+set remotevars = hostvars[host]
+conn {{ construct name based on local and remote host properties }}
+    left={{ inventory_hostname }}
+    leftid={{ vpn_hostid }}
+    right={{ host }}
+    rightid={{ remotevars.remoteid }}
+```
+
+For the host-to-host cases where we need to specify the relationships explicitly, we
```

Comment:
@simo5 did I capture the gist of what we were discussing?

Reply:
I am not sure, my idea was to have something like this:
- vpn-tunnel [some name] // like "host1-host2", could be autogenerated
  - hosts:
     - host1.my.domain
     - host2.my.domain
  - shared secret: {whatever} // this is optional of course

Then we run the playbook and if both hosts are in the inventory we configure both sides of the tunnel. If only one is we configure only that side and consider the other as "managed by a third party".

If you have more than 2 hosts then you do a full mesh configuration among them, examples:

Full mesh done by configuring individual tunnels:
- vpn-tunnel "all-group1"
  - type: host-to-host
  - hosts:
    - host1.group1.mydomain
    - host2.group1.mydomain
    - host3.group1.mydomain
    - host4.group1.mydomain
  - secret-type: keypair

For each of these hosts in the playbook inventory, the role would (in 3 rounds of connecting back and forth to each host):
- create a key pair and pull the public key into the admin host
- configure 3 tunnels to each of the other hosts using the public keys retrieved from each of the hosts
- test that each connection came up properly

Full mesh done via actual mesh and certs:
- vpn-tunnel "all-group1"
  - type: mesh
  - hosts:
    - host1.group1.mydomain
    - host2.group1.mydomain
    - host3.group1.mydomain
    - host4.group1.mydomain
  - secret-type: certificate
  - ca-certificate: blah
  - name constraints: blah // or whatever other parameter makes sense to match certs

This one would configure each host to accept connections from other hosts with a certificate signed by the provided CA and matching some name restriction (or potentially the whole list of hosts is add to a permit-list and SAN names are checked, whatever works for IPSec.

The main feature is that no matter what underlying technology or topology the declaration of the "vpn-tunnel" is always the same, and only specific properties change to steer the role to apply one or another config.
It also avoids to have multiple configurations for each host you want to add in that need to be kept in sync whenever a connection options needs to be changed (say you want to start excluding a specific cipher or host), instead you have a single configuration even for large groups of host-to-host tunnels.